src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,247 @@
+            Content = new StreamContent(_responseStream);
+        }
+    }
+
+        internal sealed class CurlResponseStream : Stream
+        {
+            private volatile bool _disposed = false;
+            private Stream _innerStream;
+            private readonly object _lockObject = new object();
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+            private readonly ManualResetEventSlim _dataRequestedEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _dataReadyEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _noMoreDataEvent = new ManualResetEventSlim(false);
+
+            internal CurlResponseStream() : base()
+            {
+            }
+
+            internal void SignalComplete(bool isCancellation)
+            {
+                if (!_disposed)
+                {
+                    _cancellationTokenSource.Cancel();
+                }
+            }
+
+            internal unsafe void WaitAndSignalReaders(byte* pointer, long length)
+            {
+                CheckDisposed();
+                _dataRequestedEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);
+                lock (_lockObject)
+                {
+                    _innerStream = new UnmanagedMemoryStream(pointer, length);
+                    _noMoreDataEvent.Reset();
+                    _dataRequestedEvent.Reset();
+                    _dataReadyEvent.Set();
+                }
+                _noMoreDataEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);;
+            }
+
+            public override bool CanRead
+            {
+                get
+                {
+                    return !_disposed;
+                }
+            }
+
+            public override bool CanSeek
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override bool CanWrite
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override long Length
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override long Position
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+
+                set
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override void Flush()
+            {
+                // Nothing to do.
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
```

Comment:
(I should also mention there's another handy optimization that's useful sometimes.  AsyncTaskMethodBuilder is like TaskCompletionSource and is what the C# compiler uses to back an async method. You construct one, and it has a Task property just like TaskCompletionSource, and a SetResult and SetException method just like TCS.  The main difference is that AsyncTaskMethodBuilder is a struct, so you avoid the extra allocation for the TCS itself.   It also lacks a SetCanceled method, instead translating OperationCanceledExceptions passed to SetException to mean that the task should end as Canceled.  You also have to make sure to access the Task property before there's any chance that concurrent code could call Set*, e.g. by accessing Task in the wrapping type's ctor... that forces it to be initialized, which is important to do promptly because it's a struct, and no synchronization is used internally during the lazy initialization of the Task.)